### ISSUE TITLE
Use argument buffers on Metal

### ISSUE THREAD

**Issue by kanerogers** [gfx-rs/wgpu#4491]
## Background
This is a companion to the wgpu issue https://github.com/gfx-rs/wgpu/issues/3334

Using [argument buffers](https://developer.apple.com/documentation/metal/buffers/about_argument_buffers) would simplify the Metal shaders emitted by naga. Instead of having to pass each binding to the shader function as an argument, each binding group can be bundled into a struct which can then be used throughout the shader. For example:

### Input wgsl
```wgsl
@group(0) @binding(0) var u_texture : texture_2d<f32>;
@group(0) @binding(1) var u_sampler : sampler;

@fragment
fn frag_main(@location(0) uv : vec2<f32>) -> @location(0) vec4<f32> {
  return textureSample(u_texture, u_sampler, uv);
}
```

### Without argument buffers
```metal
fragment frag_mainOutput frag_main(
  frag_mainInput varyings_1 [[stage_in]]
, metal::texture2d<float, metal::access::sample> u_texture [[user(fake0)]]
, metal::sampler u_sampler [[user(fake0)]]
  metal::float4 color = u_texture.sample(u_sampler, uv_1);
  return frag_mainOutput { color };
)
```

### With argument buffers
```metal
struct Group_0 {
   metal::texture2d<float, metal::access::sample> u_texture [[id(0)]];
   metal::sampler u_sampler [[id(1)]];
};

fragment frag_mainOutput frag_main(
  frag_mainInput varyings_1 [[stage_in]]
  device Group_0 & group_0 [[buffer(0)]]
) {
  metal::float4 color = group_0.u_texture.sample(group_0.u_sampler, uv_1);
  return frag_mainOutput { color };
}
```

**Comment by kanerogers** [comment]
Some updates on this one:

- I've been struggling to get a solid week blocked off to focus on this issue, so my output has been a little sporadic to date
- With a hacky, MVP implementation I've been able to get the "hello-cube" example working with wgpu/naga
- My core struggle has been figuring out how much to keep / how much to throw away with the existing implementation as I'm wary of modifying a large, complex codebase that I'm not familiar with

Some notes (for my reference, or for anyone who wants to pick this up should I get hit by a bus):

- I am *reasonably confident* that a 1:1 mapping can be made between a WGPU bind group and an argument buffer. That is, the four kinds of **resources** defined in [the WSGL spec](https://www.w3.org/TR/WGSL/#resource-interface) (Uniform Buffers, Storage Buffers, Texture resources and Sampler resources) are all legal members of argument buffers, per section 2.13 of [the MSL spec](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf):
> Argument buffers extend the basic buffer types to include pointers (buffers), textures, texture buffers, and samplers.

- With this in mind, I believe it makes more sense to preserve the structure of a bind group on both the wgpu and naga side to more easily populate (on the naga side) and encode (on the wgpu side) the relevant argument buffer.

**Comment by teoxoy** [comment]
Some considerations:
- argument buffers require `MSL 2` (i.e. `iOS 11.0+` `macOS 10.13+`); the WebGPU spec [is ok with this](https://github.com/gpuweb/gpuweb/issues/1069) and so is [Firefox](https://support.mozilla.org/en-US/kb/firefox-users-macos-1012-1013-1014-moving-to-extended-support) but this might still impact naga/wgpu users
- argument buffer tier 1 doesn't support writeable textures ([ref](https://developer.apple.com/documentation/metal/buffers/about_argument_buffers))

**Comment by jimblandy** [comment]
@kanerogers The "Without argument buffers" example in the top comment seems mangled - is that valid MSL?

**Comment by kanerogers** [comment]
> @kanerogers The "Without argument buffers" example in the top comment seems mangled - is that valid MSL?
> 
> 
Ah, good pick up! 😅

No, it's not valid MSL. This was just a rough sketch I wrote, mostly for myself, when I was first looking into the project.

**Comment by cwfitzgerald** [comment]
Closing as this is redundant to #3334, now that the repos are merged.

**Comment by mikialex** [comment]
Currently the maxStorageBuffersPerShaderStage on Metal is limited to 31, will the using of argument buffer unlock this limitation to a very large number? I see this issue closed because of the #3334, if the #3334 get implemented, can maxStorageBuffersPerShaderStage be raised a lot?

Background: I have shader that using more than 31 storage buffers so it can not run on Metal now and I'm not able to migrate my shader to using binding_array even if binding_array on Metal is supported.

**Comment by cwfitzgerald** [comment]
It will be a different limit (see #6738) but yes, it will be on the order of a million.

**Comment by mikialex** [comment]
So I suppose you mean the maxStorageBuffersPerShaderStage will not be raised, but there will be another new millions level limit for bind_array, which means I still need to somehow migrate my code using bindless resource to use so much binging in metal?

**Comment by cwfitzgerald** [comment]
Correct!

**Comment by teoxoy** [comment]
Reopening since https://github.com/gfx-rs/wgpu/issues/3334 was narrowed down to using argument buffers for `binding_array`s only.

**Comment by fanciful-marmot** [comment]
For clarity, is this the issue that needs to be resolved before `maxStorageBuffersPerShaderStage` can be raised above 8 for Firefox on mac?

**Comment by teoxoy** [comment]
Yes, although we can increase it to 9 (done in https://github.com/gfx-rs/wgpu/pull/9709) just fine and maybe even 10 but that requires some more changes (by merging usage of 2 buffers we use internally).

**Comment by fanciful-marmot** [comment]
Oh that would be great! Increasing it to 10 to match Chrome would actually help a lot

**Comment by oscarbg** [comment]
+1 for increasing maxStorageBuffersPerShaderStage to 10..
there is content in the wild failing because of that..
for example this demo:
https://jure.github.io/webphysics/?ref=webgpu.com

displays everywhere except firefox on Macos..

error is:
```
This demo requires 10 storage buffers per shader stage, but this adapter only supports 8.

ek@https://jure.github.io/webphysics/assets/avbdReferenceConfig-BdajnPuj.js:2055:14775
```

EDIT: also this demo also needs shader-subgroup feature so doesn't run the physhics compute kernels OK on Firefox on any OS
